### ISSUE TITLE
False positive on class generic type argument when constraint includes string in union.

### ISSUE THREAD

**Issue by alwaysmpe** [microsoft/pyright#8112]
May be easier to just read the code to get what the problem is.
A root class has a generic type argument that is constrained to a collection of one of int/float/bool/str.
When a subclass has a constrained type that is a union of the above except str, it can pass a sequence
of the generic type to the parent class without error. When the constrained type includes str there's an error (at end)

```python
from collections.abc import Sequence

type NumType = int | float | bool
type DiscreteType = str | bool

type SequenceT = Sequence[int] | Sequence[float] | Sequence[str] | Sequence[bool]

class Collection[ST: SequenceT]:
    pass

# Below does not raise an error.
class NumColl[T: NumType](Collection[Sequence[T]]):
    pass

# Below does not raise an error.
class StrColl[T: str](Collection[Sequence[T]]):
    pass

# Below raises an error
class DiscreteColl[T: DiscreteType](Collection[Sequence[T]]):
    pass
```

Error:
```
Type "Sequence[T@DiscreteColl]" cannot be assigned to type variable "ST@Collection"
  Type "Sequence[T@DiscreteColl]" is incompatible with bound type "SequenceT" for type variable "ST@Collection"
    Type "Sequence[T@DiscreteColl]" is incompatible with type "SequenceT"
      "Sequence[T@DiscreteColl]" is incompatible with "Sequence[int]"
        Type parameter "_T_co@Sequence" is covariant, but "T@DiscreteColl" is not a subtype of "int"
          Type "str* | bool*" is incompatible with type "int"
      "Sequence[T@DiscreteColl]" is incompatible with "Sequence[float]"
        Type parameter "_T_co@Sequence" is covariant, but "T@DiscreteColl" is not a subtype of "float"
          Type "str* | bool*" is incompatible with type "float"
  ...  (reportInvalidTypeArguments)
```

I would (personally) expect none of the subclasss to error. Alternatively, all should error, but not the case when the subclass is constrained to a union including string. Changing the union at the top to be a union of types that doesn't include string removes the error. Constraining the type to be a string removes the error.

Above code errors in pylance on VS Code, pyright command line and pyright playground. Pylance version: v2024.6.1

**Comment by erictraut** [comment]
Pyright's behavior here is correct. For comparison, mypy generates the same results.

First, it's important to understand that the type `float | int | bool` is equivalent to the type `float`. They both describe the same type because the annotation `float` in the Python type system implicitly includes a union with `int` (see [this section](https://typing.readthedocs.io/en/latest/spec/special-types.html#special-cases-for-float-and-complex) of the typing spec), and `bool` is a subtype of `int`. With that in mind, you can simplify the statement `type NumType = int | float | bool` to `type NumType = float`. Likewise, you can simplify the statement `type SequenceT = Sequence[int] | Sequence[float] | Sequence[str] | Sequence[bool]` to `type SequenceT = Sequence[float] | Sequence[str]`.

Second, the type `Sequence[bool] | Sequence[str]` is a subtype of `Sequence[bool | str]`, but the converse is not true. Therefore `Collection[Sequence[T]]` is not allowed if `T` has an upper bound of `str | bool` and the type variable for `Collection` has an upper bound of `Sequence[float] | Sequence[str]`.
